Add H.265 support to stream component

uvjustin opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Context: https://community.home-assistant.io/t/add-support-for-h-265-in-stream-component/205661

A lot of cameras support H.265 encoding which uses considerably lower bandwidth than H.264 encoding does. The standard for H.265 over HLS requires using fragmented mp4 (fmp4) segments instead of transport stream (TS) segments which are currently used in the stream component's HLS implementation.
This PR adds support for H.265 by adjusting regular mp4 segments to create fmp4 segments which can be used to stream either H.264 or H.265 over HLS.
However, note that H.265 is not well supported by non-Apple devices.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 22424
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

uvjustin committed:
Add H.265 support to stream component

uvjustin committed:
Change find_box to generator

hunterjm commented:
Thanks Justin. I’ve been meaning o look into this for a while now. Will review more thoroughly over the weekend.

uvjustin commented:
BTW, I was able to get H.265 working on an Android device by modifying the Frontend and the Android App to use Exoplayer for HLS when supported.

hunterjm commented on homeassistant/components/stream/hls.py:
```diff
@@ -109,14 +137,67 @@ def name(self) -> str:
     def format(self) -> str:
         """Return container format."""
-        return "mpegts"
+        return "mp4"
```
Can we not use `fmp4` to create the fragmented MP4 directly instead of during HLS rendering or is that not an option in FFMPEG out of the box?

uvjustin replied:
> 
> 
> In general this looks good. What does desktop browser support look like? I left a few comments comparing to the [hlsenc](https://github.com/FFmpeg/FFmpeg/blob/master/libavformat/hlsenc.c) provided by FFMPEG. Searching that file for `SEGMENT_TYPE_FMP4` shows the additional logic they have to support `m4s`. Obviously we aren't using that as it writes files to disk, but perhaps it's a good reference for our (simplified) implementation.

Yes, the reason I used this approach is because PyAV isn't written to deal with multiple file IO in ffmpeg.
I had to do a bit of digging into mp4 container format specifications and figure out the differences between mp4 and fmp4.
mp4 containers are set up with a hierarchy of boxes, and while most of the data is the same, there is some slight editing needed to convert regular mp4 to fmp4.
I discussed browser support in the forum post here: https://community.home-assistant.io/t/add-support-for-h-265-in-stream-component/205661 . Basically, Windows you are SOL unless you use a custom chromium or if/when Microsoft decides to add H.265 functionality back into the new Edge.
Here are a few links on mp4/fmp4 for your reference:
https://stackoverflow.com/questions/35177797/what-exactly-is-fragmented-mp4fmp4-how-is-it-different-from-normal-mp4
https://www.ramugedia.com/mp4-container
https://www.cimarronsystems.com/wp-content/uploads/2017/04/Elements-of-the-H.264-VideoAAC-Audio-MP4-Movie-v2_0.pdf
https://observablehq.com/@benjamintoofer/iso-base-media-file-format
https://www.programmersought.com/article/99832446840/
https://www.file-recovery.com/mp4-signature-format.htm

hunterjm commented on homeassistant/components/stream/hls.py:
```diff
@@ -109,14 +137,67 @@ def name(self) -> str:
+    @property
+    def container_options(self) -> dict:
+        """Return container options."""
+        return {"movflags": "empty_moov+default_base_moof"}
```
The FFMPEG HLS muxer sets the following options when writing `fmp4` segments: https://github.com/FFmpeg/FFmpeg/blob/master/libavformat/hlsenc.c#L866-L867

Any reason ours are different?

uvjustin replied:
'dash' wasn't documented, but if you have a look here https://github.com/FFmpeg/FFmpeg/blob/d5247fb1da615a7ccff9fa7271bc8aba23e0de25/libavformat/movenc.c#L6317 , it sets the same flags that we're using (empty moov+default_base_moof) plus the flag FF_MOV_FLAG_FRAGMENT. But FF_MOV_FLAG_FRAGMENT is redundant since it gets set here anyway: https://github.com/FFmpeg/FFmpeg/blob/d5247fb1da615a7ccff9fa7271bc8aba23e0de25/libavformat/movenc.c#L6311
frag_custom isn't necessary because that just specifies to cut the fmp4 manually, but since we are not using the fmp4 option this doesn't matter.
As for delay_moov, I didn't look into that. A quick google shows that it might keep the start time pegged at 0: https://stackoverflow.com/questions/44731531/how-to-keep-the-start-time-0-when-writing-a-mp4video-to-stdout-with-ffmpeg

hunterjm commented on homeassistant/components/stream/hls.py:
```diff
@@ -109,14 +137,67 @@ def name(self) -> str:
+        return {"movflags": "empty_moov+default_base_moof"}
+
+
+def find_box(segment: io.BytesIO, target_type: bytes, box_start: int = 0) -> int:
```
Can you walk me through the need for these methods? Is there a better place for them to live? Perhaps a utils file?

uvjustin replied:
This tools are needed to edit the fmp4 boxes. We could move them to another file...

T-bond commented:
> BTW, I was able to get H.265 working on an Android device by modifying the Frontend and the Android App to use Exoplayer for HLS when supported.

Will you upstream those changes too, when and if this PR gets merged?
I am looking forward using my 5 new IP camera, with H.265 support.

hunterjm reviewed (commented):
In general this looks good. What does desktop browser support look like? I left a few comments comparing to the [hlsenc](https://github.com/FFmpeg/FFmpeg/blob/master/libavformat/hlsenc.c) provided by FFMPEG. Searching that file for `SEGMENT_TYPE_FMP4` shows the additional logic they have to support `m4s`. Obviously we aren't using that as it writes files to disk, but perhaps it's a good reference for our (simplified) implementation.

uvjustin commented on homeassistant/components/stream/worker.py:
```diff
@@ -29,7 +29,15 @@ def create_stream_buffer(stream_output, video_stream, audio_frame):
+        mode="w",
+        format=stream_output.format,
+        container_options={
+            "video_track_timescale": str(int(1 / video_stream.time_base)),
```
This was important. The video_track_timescale didn't seem to get set correctly from video_stream template without it (it was using the base framerate and then calculating a timescale based off that (https://github.com/FFmpeg/FFmpeg/blob/d5247fb1da615a7ccff9fa7271bc8aba23e0de25/libavformat/movenc.c#L6498). Since we are not reencoding our streams, we should just use the timescale of the input stream.

uvjustin commented on homeassistant/components/stream/worker.py:
```diff
@@ -101,29 +112,36 @@ def stream_worker(hass, stream, quit_event):
-            # timestamp by the time base, which gets us total seconds.
-            # By then dividing by the sequence, we can calculate how long
-            # each segment is, assuming the stream starts from 0.
-            segment_duration = (packet.pts * packet.time_base) / sequence
```
Not sure why but it seems like the original calculated the average segment length instead of the individual segment length.

hunterjm replied:
Was probably just me trying stuff until it worked some years back. Since keyframe's are generally at a set interval unless the feed has a dynamic frame rate the average and individual duration would probably be the same anyway. Good catch though.

uvjustin commented:
> > BTW, I was able to get H.265 working on an Android device by modifying the Frontend and the Android App to use Exoplayer for HLS when supported.
> 
> Will you upstream those changes too, when and if this PR gets merged?

Sure. Might need some help or input on that - I'm not a professional coder and it's my first time looking at Android / Kotlin. There's a bit of work in overlaying the exoplayer view on the webview. Also the exoplayer controls look a little different than the html5 ones do. Also the "preload stream" checkbox element might have to be moved a little.

uvjustin committed:
Move fmp4 utilities to fmp4utils.py

uvjustin committed:
Add minimum segments and segment durations

uvjustin commented:
@hunterjm 
Rebased and moved the box utilities to a separate file.
Moved num_segments to MAX_SEGMENTS in .const.
Added MIN_SEGMENTS and set it to 2 so we have a little bit of a buffer when we start.
Added MIN_SEGMENT_DURATION and set it to 1.5 so each segment is at least 1.5 seconds long (many camera keyframe rates are about 1 second, so this would make each segment 2 seconds instead of the 1 second previously).
The addition of MIN_SEGMENTS AND MIN_SEGMENT_DURATION will increase latency from about 1-2 seconds to 4-5 seconds, but there should be less buffering and also less overhead in both remuxing and transport.
FWIW, Apple seems to indicate 6 segments of 6 seconds each: https://developer.apple.com/documentation/http_live_streaming/hls_authoring_specification_for_apple_devices, but that doesn't work for us and doesn't seem to be necessary.
We can always implement LL-HLS later when it becomes supported in hls.js.

uvjustin commented on homeassistant/components/stream/core.py:
```diff
@@ -78,12 +82,12 @@ def segments(self) -> List[int]:
     @property
     def target_duration(self) -> int:
-        """Return the average duration of the segments in seconds."""
+        """Return the max duration of any given segment in seconds."""
```
The target duration tag seems to be defined as the max duration

uvjustin committed:
Remove MIN_SEGMENTS

uvjustin committed:
Fix when container_options is None

uvjustin commented on homeassistant/components/stream/hls.py:
```diff
@@ -109,14 +136,19 @@ def name(self) -> str:
+    @property
+    def container_options(self) -> dict:
+        """Return container options."""
+        return {"movflags": "frag_custom+empty_moov+default_base_moof"}
```
Added frag_custom since we are no longer cutting segments on every keyframe (default frags on every keyframe)

uvjustin reviewed (commented):
Took out MIN_SEGMENTS because buffering strategy should probably be handled on the player side (although only starting after 2 segments are available seems reasonable)

uvjustin reviewed (commented):
Minor change to keep recorder working

uvjustin commented:
I did quite a bit of refactoring and added support for audio (and removed the audio stream altogether if no audio stream was on the input). I can add that to this PR or submit another PR later.

hunterjm commented:
> I did quite a bit of refactoring and added support for audio (and removed the audio stream altogether if no audio stream was on the input). I can add that to this PR or submit another PR later.

Let's make that a separate PR if it's not already in this commit. It's been really hectic here so sorry again for the delay in review/merge.

uvjustin commented:
> 
> 
> > I did quite a bit of refactoring and added support for audio (and removed the audio stream altogether if no audio stream was on the input). I can add that to this PR or submit another PR later.
> 
> Let's make that a separate PR if it's not already in this commit. It's been really hectic here so sorry again for the delay in review/merge.

Sure, no problem. I'll stash it for now.

hunterjm commented on homeassistant/components/stream/core.py:
```diff
@@ -34,21 +34,20 @@ class Segment:
 class StreamOutput:
     """Represents a stream output."""
 
-    num_segments = 3
```
Removing this breaks references in the stream record service here: https://github.com/home-assistant/core/blob/dev/homeassistant/components/stream/__init__.py#L228

uvjustin replied:
Thanks, didn't catch that.

hunterjm reviewed (changes requested):
So far this looks really good. I've tested it locally and only found one small issue. If you wouldn't mind also going through and updating/removing tests for these changes we can go ahead and get this merged!

uvjustin committed:
Fix missing num_segments and update tests

uvjustin commented on tests/components/stream/test_hls.py:
```diff
@@ -99,15 +106,16 @@ async def test_stream_ended(hass):
     # Run it dead
-    segments = 0
-    while await track.recv() is not None:
-        segments += 1
```
track.recv() reads from the end of the queue, so segments might get skipped

uvjustin committed:
Remove unnecessary mock attribute

hunterjm reviewed (changes requested):
Sorry to remove approval, but it looks like the record service is still having issues. I'm attempting to research further.

uvjustin commented:
> 
> 
> Sorry to remove approval, but it looks like the record service is still having issues. I'm attempting to research further.

Interesting, let me know what you find.

uvjustin committed:
Fix Segment construction in test_recorder_save

hunterjm commented:
> Interesting, let me know what you find.

So it looks like the pts of each segment is now being reset to `0` for an initial value. This causes issues in `recorder_save_worker` the way it's currently implemented because when we attempt to mux multiple segments together the pts and dts of the next segment is incorrect in the new context of the single file. We need to add logic to modify the pts/dts of segments 2-n when muxing together in order to re-create a single file stream.

uvjustin commented:
> 
> 
> > Interesting, let me know what you find.
> 
> So it looks like the pts of each segment is now being reset to `0` for an initial value. This causes issues in `recorder_save_worker` the way it's currently implemented because when we attempt to mux multiple segments together the pts and dts of the next segment is incorrect in the new context of the single file. We need to add logic to modify the pts/dts of segments 2-n when muxing together in order to re-create a single file stream.
> 
> edit: interestingly enough, I can reproduce it consistantly, but it appears to only be the first two segments with this issue.

Hmm, I've been using my significantly refactored version of the component that also works with audio. When I was working on it I realized an issue with initial packets - the first packet dts/pts we get is indeed wrong, and we shouldn't be using it at all. See https://trac.ffmpeg.org/ticket/5018 . Essentially you get an adjuster with a PTS*time_base of about 1/4 to 1/2 a second, and so you end up discarding all the packets from the first frame until that point. That might be enough to mess up the final muxing. I ended up fixing this in the subsequent version by peeking ahead to the second packet to get the correct DTS.
For this PR I can just set the first_pts to 0 and that should be good enough for most?

hunterjm commented:
> > > Interesting, let me know what you find.
> > 
> > 
> > So it looks like the pts of each segment is now being reset to `0` for an initial value. This causes issues in `recorder_save_worker` the way it's currently implemented because when we attempt to mux multiple segments together the pts and dts of the next segment is incorrect in the new context of the single file. We need to add logic to modify the pts/dts of segments 2-n when muxing together in order to re-create a single file stream.
> > edit: interestingly enough, I can reproduce it consistantly, but it appears to only be the first two segments with this issue.
> 
> Hmm, I've been using my significantly refactored version of the component that also works with audio. When I was working on it I realized an issue with initial packets - the first packet dts/pts we get is indeed wrong, and we shouldn't be using it at all. See https://trac.ffmpeg.org/ticket/5018 . Essentially you get an adjuster with a PTS*time_base of about 1/4 to 1/2 a second, and so you end up discarding all the packets from the first frame until that point. That might be enough to mess up the final muxing. I ended up fixing this in the subsequent version by peeking ahead to the second packet to get the correct DTS.
> For this PR I can just set the first_pts to 0 and that should be good enough for most?

Does that refactored component work with the record service? Are you able to checkout this branch and reproduce the issue I described by calling camera.record? I'm using `duration: 20` and `lookback: 5`.

uvjustin commented:
> 
> 
> > > > Interesting, let me know what you find.
> > > 
> > > 
> > > So it looks like the pts of each segment is now being reset to `0` for an initial value. This causes issues in `recorder_save_worker` the way it's currently implemented because when we attempt to mux multiple segments together the pts and dts of the next segment is incorrect in the new context of the single file. We need to add logic to modify the pts/dts of segments 2-n when muxing together in order to re-create a single file stream.
> > > edit: interestingly enough, I can reproduce it consistantly, but it appears to only be the first two segments with this issue.
> > 
> > 
> > Hmm, I've been using my significantly refactored version of the component that also works with audio. When I was working on it I realized an issue with initial packets - the first packet dts/pts we get is indeed wrong, and we shouldn't be using it at all. See https://trac.ffmpeg.org/ticket/5018 . Essentially you get an adjuster with a PTS*time_base of about 1/4 to 1/2 a second, and so you end up discarding all the packets from the first frame until that point. That might be enough to mess up the final muxing. I ended up fixing this in the subsequent version by peeking ahead to the second packet to get the correct DTS.
> > For this PR I can just set the first_pts to 0 and that should be good enough for most?
> 
> Does that refactored component work with the record service? Are you able to checkout this branch and reproduce the issue I described by calling camera.record? I'm using `duration: 20` and `lookback: 5`.

Yes, record works, but I did not test it extensively.
Sure, I can checkout this branch and test that out, but I'll have to do it a bit later.

hunterjm commented:
> Yes, record works, but I did not test it extensively.
> Sure, I can checkout this branch and test that out, but I'll have to do it a bit later.

This appears to be triggered with the `lookback` segments only. Works fine if no `lookback` is defined.

uvjustin commented on homeassistant/components/stream/__init__.py:
```diff
@@ -225,7 +226,7 @@ async def async_handle_record_service(hass, call):
     hls = stream.outputs.get("hls")
     if lookback > 0 and hls:
-        num_segments = min(int(lookback // hls.target_duration), hls.num_segments)
+        num_segments = min(int(lookback // hls.target_duration), MAX_SEGMENTS)
```
Could this be the problem? If there are no segments ready yet, hls.target_duration gives a 0 and this will blow up

hunterjm replied:
That's not the issue. I am getting segments back, but when the recorder worker opens them, the pts/dts of the packets start at 0 instead of `segment.start_pts[0]` like they do with every other segment after that.

hunterjm replied:
I really can't see where your changes affect that though, unless my hijacking of the `hls` plugin for use in `recorder` is grabbing segments after they've already been modified by the `m4s` helpers.

hunterjm committed:
fix recorder with lookback

hunterjm committed:
Merge branch 'h265' of https://github.com/uvjustin/home-assistant into h265

hunterjm commented:
Pushed a change with the recorder fix for lookback. Once tests pass LGTM, then open up the audio PR and see if we can't get the best of both worlds in 0.115!

uvjustin commented on homeassistant/components/stream/recorder.py:
```diff
@@ -36,9 +36,9 @@ def recorder_save_worker(file_out: str, segments: List[Segment]):
-                if first_pts is None:
-                    first_pts = packet.pts
-
+                if packet.pts < segment.start_pts[0]:
```
Does this just work around when segments[0].start_pts[0] is incorrect?

hunterjm replied:
It gets around when `packet.pts` is `0` based instead of `start_pts` based, which seems to happen for the "lookback" segments we pull from the hls provider.

uvjustin replied:
Interesting...can't see why the lookback segments would be coming out as 0...will have to take a look with fresh eyes later.

uvjustin replied:
The PTS and DTS get set to zero when the container option "empty_moov" is set. I think it's from this line https://ffmpeg.org/doxygen/trunk/movenc_8c_source.html#l05657
The timestamps get reset to match the last timestamps of the existing track, which is empty. This explains why we have to add in the tfdt offsets. From looking at that file, it seems like we actually might be able to skip the box editing if we set "frag_discont" and "fragment_index". I'll play around with it later when I have time.
BTW, any thoughts on how to deal with timestamp overflow/rollover?
